torch/functional.py

```diff
@@ -113,3 +118,91 @@ def btriunpack(LU_data, LU_pivots, unpack_data=True, unpack_pivots=True):
+        tensor2 (Tensor): Second tensor to be multiplied
+        out (Tensor, optional): Output tensor
+    """
+
+    dim_tensor1 = tensor1.dim()
+    try:
+        dim_tensor2 = tensor2.dim()
+    except AttributeError:  # not a tensor
+        return NotImplemented
+    if dim_tensor1 == 1 and dim_tensor2 == 1:
+        if out is None:
+            return torch.dot(tensor1, tensor2)
+        else:
+            raise ValueError("out must be None for 1-d tensor matmul, returns a scalar")
+    if dim_tensor1 == 2 and dim_tensor2 == 1:
+        if out is None:
+            return torch.mv(tensor1, tensor2)
+        else:
+            return torch.mv(tensor1, tensor2, out=out)
+    if dim_tensor1 == 1 and dim_tensor2 == 2:
+        if out is None:
+            return torch.mm(tensor1.unsqueeze(0), tensor2).squeeze(0)
+        else:
+            return torch.mm(tensor1.unsqueeze(0), tensor2, out=out).squeeze_(0)
+    elif dim_tensor1 == 2 and dim_tensor2 == 2:
+        if out is None:
+            return torch.mm(tensor1, tensor2)
+        else:
+            return torch.mm(tensor1, tensor2, out=out)
+    elif dim_tensor1 >= 2 and dim_tensor2 >= 2:
+        # ensure each tensor is at least 3-dimensional
+        tensor1_exp_size = torch.Size((1,) * max(3 - tensor1.dim(), 0) + tensor1.size())
+        tensor2_exp_size = torch.Size((1,) * max(3 - tensor2.dim(), 0) + tensor2.size())
+
+        # expand the batch portion (i.e. cut off matrix dimensions and expand rest)
+        expand_batch_portion = torch._C._infer_size(tensor1_exp_size[:-2], tensor2_exp_size[:-2])
+
+        # flatten expanded batches
+        tensor1_expanded = tensor1.expand(*(expand_batch_portion + tensor1_exp_size[-2:])) \
+            .contiguous().view(reduce(mul, expand_batch_portion), *tensor1_exp_size[-2:])
+        tensor2_expanded = tensor2.expand(*(expand_batch_portion + tensor2_exp_size[-2:])) \
+            .contiguous().view(reduce(mul, expand_batch_portion), *tensor2_exp_size[-2:])
+
+        # reshape batches back into result
+        total_expansion = expand_batch_portion + (tensor1_exp_size[-2], tensor2_exp_size[-1])
+        if out is None:
+            return torch.bmm(tensor1_expanded, tensor2_expanded).view(*(total_expansion))
+        else:
+            # We can only safely reshape the output if the output (after the torch.bmm call)
+            # is contiguous.  This will happen only if:
+            # 1) We force it to be contiguous
+            # 2) The output came in as contiguous
+            # 3) The output came in as the wrong size (so was resized in the torch.bmm call).
+            #
+            # But 1) is inconsistent with other functions (e.g. torch.bmm) that will maintain
+            # output non-contiguity if the size is correct (perhaps we should change this globally?)
+            # And 3) is a surprising output to accept if we aren't accepting 1).
+            # So let's just force accepting contiguous tensors.
```

Comment:
by force it to be contiguous, you mean call contiguous(), not raise an error if it's not contiguous, right?